R/data.table.R

```diff
@@ -968,7 +968,39 @@ replace_dot_alias = function(e) {
         }
         # fix for long standing FR/bug, #495 and #484
         allcols = c(names_x, xdotprefix, names_i, idotprefix)
-        if ( length(non_sdvars <- setdiff(intersect(av, allcols), c(bynames, ansvars))) ) {
+        non_sdvars <- setdiff(intersect(av, allcols), c(bynames, ansvars))
+        
+        # added 'mget' - fix for #994
+        if (any(c("get", "mget") %chin% av)){
+          # get('varname') is too difficult to detect which columns are used in general
+          # eval(macro) column names are detected via the  if jsub[[1]]==eval switch earlier above.
+          b_warn = FALSE 
+          if (is.call(jsub) && length(jsub[[1L]]) == 1L && jsub[[1L]] == ":="){
+            if (is.symbol(jsub[[2L]])) {
+              jsub_lhs_symbol = as.character(jsub[[2L]])
+              if (jsub_lhs_symbol %chin% non_sdvars) {
+                sdvars = setdiff(sdvars, jsub_lhs_symbol)
+              }
+            } else {
+              # fix for #4089 - if all names are already accounted for, we don't need to re-extract the names for the (m)get call
+              b_warn = is.null(names_i) && all(names_x %chin% c(ansvars, allbyvars))
+            }
+          }
+          if (!b_warn) {
+            if (verbose) 
+              cat("'(m)get' found in j. ansvars being set to all columns. Use .SDcols or a single j=eval(macro) instead. Both will detect the columns used which is important for efficiency.\nOld:", paste(ansvars,collapse=","),"\n")
+            ansvars = setdiff(allcols, bynames) # fix for bug #5443
+            non_sdvars = setdiff(ansvars, sdvars)
+            ansvals = chmatch(ansvars, names_x)
+            if (verbose) 
+              cat("New:",paste(ansvars,collapse=","),"\n")
+          } else { ## TODO remove this else statement in Jan. 2023
+            newvars = setdiff(allcols, bynames) # checking if old cols c("y", "x") became c("x", "y")
+            if (length(newvars) == length(ansvars) && !all(newvars == ansvars)) {
+              warning("(m)get found in j with update assignment (i.e., `(cols) := ...`). j column order was re-ordered prior to version 1.12.9. See Issue #4089")
```

Comment:
best to refer to even version numbers in messages

Reply:
Do we need to warn here? Since we're not changing any tests it could be considered as relying on untested behavior.

If we're going to warn here, there should be a way for users to turn it off. Properly I guess we'd have `options(datatable.warn.mget.reorder = TRUE)` for the next release, then change it `FALSE` the following release, remove the option the release after that?

Reply:
On re-read, this is what Jan was talking about in the comment on the main post:

https://github.com/Rdatatable/data.table/pull/4212#pullrequestreview-350862722

Reply:
I removed the warning.